azure-iot-device/tests/iothub/pipeline/test_iothub_pipeline.py

```diff
@@ -120,66 +130,89 @@ def test_pipeline_configuration(self, auth_provider):
         "Runs a SetAuthProviderOperation with the provided AuthenticationProvider on the pipeline, if using SAS based authentication"
     )
     def test_sas_auth(self, mocker, device_connection_string):
-        mocker.patch.object(pipeline_stages_base, "PipelineRootStage")
+        mocker.spy(pipeline_stages_base.PipelineRootStage, "run_op")
         auth_provider = SymmetricKeyAuthenticationProvider.parse(device_connection_string)
         pipeline = IoTHubPipeline(auth_provider)
-        op = pipeline._pipeline.run_op.call_args[0][0]
+        op = pipeline._pipeline.run_op.call_args[0][1]
```

Comment:
Good question.  This took me a while to reason my way around.  It has to do with unittest.mock's lazy attributes, and I can understand it, but I'm not sure if I can explain it well.  In the old code, because of lazy attributes, run_op gets created as an instance attribute when it's called. so the call only sees the second parameter (op),  In the new code, the spy gets added to the "pre-bound" run_op function.  Instantiation of a PipelineRoot binds the wrapped run_op function, so the spy sees both parameters (self and op).  

To explain differently:

In the old code, when you say `mocker.patch.object(pipeline_stages_base, "PipelineRootStage)`, you are telling the mock library to replace all PipelineRootStage instaces with MagicMock instances.   When the code under test calls `root=PipelineRootStage()`, root is returned as a MagicMock object with no attributes.  Then, when it calls `root.run_op(op)`, `root.run_op` gets set to a new MagicMock instance .  This run_op MagicMock is an attribute on the root instance, so it's effectively bound to the `root` object.  When the code under test calls `root.run_op(op)` it's getting the root.run_op MagicMock object and calling it with a single parameter, hence the [0][0] in the old code.  

In the new code, however, we are not replacing PipelineRootStage with a MagicMock so this lazy attribute crap doesn't come into the picture.  Instead, we are putting a wrapper around the existing PipelineRootStage.run_op implementation.    PipelineRootStage.run_op is a `pre-bound` function that takes 2 parameters -- self and op.  When the code under test calls `root=PipelineRootStage()`, it creates a PipelineRootStage object which has a wrapper around PipelineRootStage.run_op(self, op).  I'm not 100% sure how instance binding works yet, but this PipelineRootStage.run_op(self, op) object gets bound to the `root` object.  When the code under test calls `root.run_op(op)`, it binds the first parameter to `root` and the second parameter gets set to `op`.  The MagicMock (spy) object sees both parameters., so we need to look at [0][1] instead.  

Not a good explanation, but the best I can do right now.

---
In reply to: [318847028](https://github.com/Azure/azure-iot-sdk-python-preview/pull/174#discussion_r318847028) [](ancestors = 318847028)